src/ngx_http_lua_pipe.c

```diff
@@ -0,0 +1,2476 @@
+static ngx_int_t ngx_http_lua_pipe_resume(ngx_http_request_t *r);
+static void ngx_http_lua_pipe_dummy_event_handler(ngx_event_t *ev);
+static void ngx_http_lua_pipe_cleanup_helper(
+    ngx_http_lua_co_ctx_t *wait_co_ctx);
+static void ngx_http_lua_pipe_proc_read_stdout_cleanup(void *data);
+static void ngx_http_lua_pipe_proc_read_stderr_cleanup(void *data);
+static void ngx_http_lua_pipe_proc_write_cleanup(void *data);
+static void ngx_http_lua_pipe_proc_wait_cleanup(void *data);
+
+
+static ngx_rbtree_t       ngx_http_lua_pipe_rbtree;
+static ngx_rbtree_node_t  ngx_http_lua_pipe_proc_sentinel;
+
+
+#if (NGX_HTTP_LUA_HAVE_SIGNALFD)
+static int                                ngx_http_lua_signalfd;
+static struct signalfd_siginfo            ngx_http_lua_pipe_notification;
+
+#define ngx_http_lua_read_sigfd           ngx_http_lua_signalfd
+
+#else
+static int                                ngx_http_lua_sigchldfd[2];
+static u_char                             ngx_http_lua_pipe_notification[1];
+
+#define ngx_http_lua_read_sigfd           ngx_http_lua_sigchldfd[0]
+#define ngx_http_lua_write_sigfd          ngx_http_lua_sigchldfd[1]
+#endif
+
+
+static ngx_connection_t                  *ngx_http_lua_sigfd_conn = NULL;
+
+
+#define SIG(name)  { (SIG##name),  "SIG"#name }
+
+
+/* signals below are handled specially in Nginx.
+ * We need to reset them for the spawned child processes */
+ngx_http_lua_pipe_signal_t ngx_signals[] = {
+    { SIGSYS, "SIGSYS" },
+    { SIGPIPE, "SIGPIPE" },
+    { 0, NULL }
+};
+
+
+enum {
+    PIPE_ERR_CLOSED = 1,
+    PIPE_ERR_SYSCALL,
+    PIPE_ERR_NOMEM,
+    PIPE_ERR_TIMEOUT,
+    PIPE_ERR_ADD_READ_EV,
+    PIPE_ERR_ADD_WRITE_EV
+};
+
+
+enum {
+    PIPE_READ_ALL,
+    PIPE_READ_BYTES,
+    PIPE_READ_LINE,
+    PIPE_READ_ANY
+};
+
+
+#define REASON_EXIT         "exit"
+#define REASON_SIGNAL       "signal"
+#define REASON_UNKNOWN      "unknown"
+
+#define REASON_RUNNING_CODE  0
+#define REASON_EXIT_CODE     1
+#define REASON_SIGNAL_CODE   2
+#define REASON_UNKNOWN_CODE  3
+
+
+void
+ngx_http_lua_pipe_init(void)
+{
+    ngx_rbtree_init(&ngx_http_lua_pipe_rbtree,
+                    &ngx_http_lua_pipe_proc_sentinel, ngx_rbtree_insert_value);
+}
+
+
+ngx_int_t
+ngx_http_lua_pipe_add_signal_handler(ngx_cycle_t *cycle)
+{
+    int                  rc;
+    ngx_event_t         *rev;
+#if (NGX_HTTP_LUA_HAVE_SIGNALFD)
+    sigset_t             set;
+
+#else
+    struct sigaction     sa;
+#endif
+
+#if (NGX_HTTP_LUA_HAVE_SIGNALFD)
+    rc = sigemptyset(&set);
+    if (rc != 0) {
+        ngx_log_error(NGX_LOG_ERR, cycle->log, ngx_errno,
+                      "lua pipe init signal set failed");
+        return NGX_ERROR;
+    }
+
+    rc = sigaddset(&set, SIGCHLD);
+    if (rc != 0) {
+        ngx_log_error(NGX_LOG_ERR, cycle->log, ngx_errno,
+                      "lua pipe add SIGCHLD to signal set failed");
+        return NGX_ERROR;
+    }
+
+    rc = sigprocmask(SIG_BLOCK, &set, NULL);
```

Comment:
Since we are blocking `SIGCHLD` (or overriding its signal handler when we don't have `signalfd`), how come we are not respecting Nginx's original handling of the signal? https://github.com/nginx/nginx/blob/release-1.13.6/src/os/unix/ngx_process.c#L401

Reply:
@thibaultcha 
Nginx's original handling only works in the master process, and we block `SIGCHLD` in the worker process. What we lose is only the NOTICE level error message: `signal %d (%s) received from xxx`. We could add the error message in our handler.

Reply:
Yeah, worker processes usually don’t have much to reap... I did have this NOTICE log in mind though when writing the above comment (after seeing it so much when working on the SA_RESTART fix).

Reply:
@thibaultcha 
I prefer adding it in `ngx_http_lua_pipe_sigchld_event_handler`, since the signal handler will trigger this event handler eventually, so that we could avoid repeating the notice log.